File: C++/0151. 熊猫钓鱼.cpp
// we have defined the necessary header files here for this problem.
// If additional header files are needed in your program, please import here.
#include <vector>
#include <algorithm>
int main()
{  
  // please define the C++ input here. For example: int a,b; cin>>a>>b;;  
  // please finish the function body here.  
  // please define the C++ output here. For example:cout<<____<<endl; 
    int n;
    cin>>n;
    vector<vector<int>> arr;
    int m;
    
    for(int i = 0; i < 2; ++i) {
        vector<int> temp;
        for(int j = 0; j < n; ++j) {
            cin>>m;
            temp.push_back(m);
        }
        arr.push_back(temp);
    }
    
    int r = 0;
    vector<int> desk;
    
    while(!arr[0].empty() && !arr[1].empty()) {
        for(int i = 0; i < 2; ++i) {
            m = arr[i][0];
            arr[i].erase(arr[i].begin());
            auto it = find(desk.begin(), desk.end(), m);
            if(it != desk.end()) {                
                arr[i].push_back(m);
                for(auto j = desk.end() - 1; j >= it; --j) {
                    arr[i].push_back(*j);
                }
                desk.erase(it, desk.end());
            } else {
                desk.push_back(m);
            }
        }
        r++;
    }
    
    for(int i = 0; i < 2; ++i) {
        if(arr[i].empty()){
            cout<<r<<endl;
            cout<<(i == 1 ? "mm" : "xx") << endl;
            for(int j = 0; j < arr[1 - i].size() - 1; ++j) {
                cout<<arr[1 - i][j] << " ";
            }
            cout<<arr[1 - i].back()<<endl;
            return 0;
            }
    }
    return 0;
}

File: Python/0151. 熊猫钓鱼.py
# If you need to import additional packages or classes, please import here.

def func():
    # please define the python3 input here. 
    # For example: a,b = map(int, input().strip().split())
    # please finish the function body here.
    # please define the python3 output here. For example: print().
    n = int(input())
    arr = []
    arr.append(list(map(int, input().split())))
    arr.append(list(map(int, input().split())))
    p = 0
    desk = []
    r = 0
    while arr[0] and arr[1]:
        for p in range(2):
            m = arr[p].pop(0)
            if m in desk:          
                arr[p].append(m)
                i = desk.index(m)
                arr[p] += reversed(desk[i:])

                desk = desk[:i]
            else:
                desk.append(m)        
        p = 0
        r += 1
    for p in range(2):
        if not arr[p]:
            print(r)
            print("mm" if p == 1 else "xx")
            print(*arr[1 - p])
            return
if __name__ == "__main__":
    func()
